### ISSUE TITLE
Simplify the Type representation and reduce its heap usage

### ISSUE THREAD

**Issue by Marwes** [Marwes/gluon#50]
Currently there are quite a lot of different representation in the [Type enum](https://github.com/Marwes/gluon/blob/d1aaee3148d8954ccf37a1dd12d47c97a9597fd4/base/src/types.rs#L150-L181) this leads to complications in unification (and elsewhere) as types that should be equivalent can have multiple representations. To some extent these different representations are necessary either for convenience or for efficiency (`Function(Int, Float)` is a lot nicer to work with than `App(App(->, Int), Float)`).

Currently the `App` variant only exists to avoid allocation a one element vector for single argument constructors which are very common but with [smallvec](https://crates.io/crates/smallvec) (or equivalent) the `App` variant could be completely superseded by `Data`.

Other simplifications may be to replace`Function(a, b)` with `Data(-> [a, b])` and `Array(a)` with `Data([], [a]`.

Marking this as beginner as this change (though maybe a bit tedious) should just involve replacing all uses of `Type::App(a, b)` with `Type::Data(a, [b])`.

**Comment by bjz** [comment]
Here is a checklist:

- [x]: Remove `Type::App` (#52)
- [ ]: Remove `Type::Function` (#55)
- [ ]: Remove `Type::Array`

(I would put it in the issue description, but I would need to be a contributor! hint hint... 😉)

**Comment by Marwes** [comment]
Made you a contributor now, "With great power comes great responsibility" and all that :smile: .

**Comment by bjz** [comment]
Hah, of course! ^_^

**Comment by Marwes** [comment]
Also enabled the option that all forces travis checks to succeed before it being possible to merge but that might be a bad idea in case spurious errors (skeptic has been acting up the last week or so). Are you able to restart travis builds?

**Comment by bjz** [comment]
Good idea. And yes, after force-syncing my travis account, I have a retry icon on builds now.

**Comment by bjz** [comment]
Should `Array` be a `BuiltinType`?

**Comment by Marwes** [comment]
Yes. I expect partially applying array does not work now then either. It may be better to remove the `[a]` syntax for array types as I am not sure the extra syntax carries its own weight.

**Comment by bjz** [comment]
Are arrays even used anywhere? I guess I just haven't seen them in the tests...

**Comment by bjz** [comment]
Yeah, I would prefer not to have a special syntax - what about just `Array a`?

**Comment by Marwes** [comment]
There exists a few array tests in `tests/vm.rs` but thats about it.

`Array a` was what I was going to suggest make sure you update the `from_str` and `to_str` on `BuiltinType` if you change it.

**Comment by Marwes** [comment]
The `BuiltinType` enum could be removed though it would require some refactoring to make it possible (and efficient).

**Comment by bjz** [comment]
So would you then register the builtin types into the default context, with their respective kinds?

**Comment by Marwes** [comment]
I was thinking that [PrimitiveEnv](https://github.com/Marwes/gluon/blob/9058c253c23358cdfc249e88c5789bd3b4f6e98c/base/src/types.rs#L115-L117) could be extended to provide all primitive types as well as [the functions](https://github.com/Marwes/gluon/blob/9058c253c23358cdfc249e88c5789bd3b4f6e98c/base/src/types.rs#L470-L552) that are now on `Type` (all of those functions may not be needed).

**Comment by Marwes** [comment]
That change to `PrimitiveEnv` would be useful regardless of whether `BuiltinType` is kept or removed #41.

**Comment by brendanzab** [comment]
So [`<Typed for Expr<Id>>::env_type_of`](https://github.com/Marwes/gluon/blob/9058c253c23358cdfc249e88c5789bd3b4f6e98c/base/src/ast.rs#L579-L587) requires the primitive types. Do you think we could just move the getters into `TypeEnv`, and get rid of `PrimitiveEnv`?

**Comment by brendanzab** [comment]
I just wouldn't know what to return for `<TypeEnv for ()>::{float, ..}`...

**Comment by Marwes** [comment]
It might be better to require `PrimitiveEnv` in `env_type_of`. Not all `TypeEnv` needs to be `PrimitiveEnv`such as the tuple implementation for `PrimitiveEnv`

**Comment by Marwes** [comment]
Ah ok. In that case maybe the best long term improvement is to remove the `type_of` method. It might be a bit annoying to thread the `PrimitiveEnv` everywhere it is needed however.

The other option may be to implement `PrimitiveEnv` for `()` which should basically be the implementations for the static methods which currently exist on `Type`. In that case I think `PrimitiveEnv` and `TypeEnv` might as well be merged as well since the 'static methods' could just be default implementations on `TypeEnv`

(Only the check crate seem to use it)
```
$ grep "\.type_of" check/src/*.rs
check/src/completion.rs:        self.0 = Some(expr.type_of());
check/src/completion.rs:        self.0 = Some(ident.type_of());
check/src/completion.rs:            let typ = expr.type_of();
check/src/rename.rs:                            for (typ, arg) in types::arg_iter(&bind.type_of())
check/src/typecheck.rs:                self.unify_span(bind.name.span(), &bind.type_of().clone(), typ);
check/src/typecheck.rs:                            match self.tc.type_of_alias(alias, args) {
```

**Comment by brendanzab** [comment]
What would `()` return for `PrimitiveEnv::get_bool`? Also, do you mind if I switch impling on `()` to a `EmptyTypeEnv` unit struct?

**Comment by Marwes** [comment]
Not sure actually, maybe we have to remove `type_of` instead and not have a `EmptyTypeEnv: PrimitiveEnv`. Or maybe the `function`, `int`, `string`, etc methods should just be on `TypeEnv` instead leaving `PrimitiveEnv` alone?

**Comment by brendanzab** [comment]
I'll try removing `type_of` and see what breaks!

**Comment by brendanzab** [comment]
I'm also thinking of switching to using static dispatch for `env_type_of`, which is more idiomatic in most Rust code. Is that ok? Ie.

```diff
-    fn env_type_of(&self, env: &TypeEnv) -> ASTType<Self::Id>;
+    fn env_type_of<E: TypeEnv>(&self, env: &E) -> ASTType<Self::Id>;
```

**Comment by brendanzab** [comment]
Any idea why in https://github.com/Marwes/gluon/blob/master/check/src/typecheck.rs#L890 we're using `bind.type_of()` instead of `bind.env_type_of(&self.environment)`?

Eg.

```diff
            for (typ, bind) in types.into_iter().zip(bindings.iter_mut()) {
                // Merge the variable we bound to the name and the type inferred
                // in the expression
-               self.unify_span(bind.name.span(), &bind.type_of().clone(), typ);
+               let other_bind = bind.env_type_of(&self.environment);
+               self.unify_span(bind.name.span(), &other_bind, typ);
            }
```

**Comment by brendanzab** [comment]
The only other places I could find that need `type_of` but don't have an environment in scope are:

- https://github.com/Marwes/gluon/blob/a59692f8f98cab6afaf5a3b0f33c81134a9b37de/check/src/completion.rs#L31-L36
- https://github.com/Marwes/gluon/blob/a59692f8f98cab6afaf5a3b0f33c81134a9b37de/check/src/completion.rs#L86
- https://github.com/Marwes/gluon/blob/a59692f8f98cab6afaf5a3b0f33c81134a9b37de/check/tests/pass.rs#L140
- https://github.com/Marwes/gluon/blob/a59692f8f98cab6afaf5a3b0f33c81134a9b37de/check/tests/pass.rs#L197-L201

Changing the others to use the envs in scope didn't seem to make the tests fail. Any idea how to switch those out?

**Comment by Marwes** [comment]
> Any idea why in https://github.com/Marwes/gluon/blob/master/check/src/typecheck.rs#L890 we're using bind.type_of() instead of bind.env_type_of(&self.environment)?

Probably I forgot to change it as `type_of` was added first `env_type_of` came later. Might be it wasn't needed however (but it won't hurt to have it).

**Comment by Marwes** [comment]
For the completion file just thread the state in `OnFound` and `Suggest`. This will require you to change the [completion test code](https://github.com/Marwes/gluon/blob/a59692f8f98cab6afaf5a3b0f33c81134a9b37de/check/tests/completion.rs) as well but you can use [MockEnv](https://github.com/Marwes/gluon/blob/a59692f8f98cab6afaf5a3b0f33c81134a9b37de/check/tests/support/mod.rs#L51) as a `PrimitiveEnv`.

**Comment by Marwes** [comment]
Taking a look at adding `smallvec`. Looks to be some rough edges in the library so I think I will fix a couple things and submit a PR to the `smallvec` repo first.
